### PR TITLE
Add field validation on structs

### LINKED ISSUE #43: Bug: Field constraints not evaluated on structs
This project looks really interesting, and I look forward to seeing how it develops. The struct support seems quite good, but it doesn't seem to support field constraints at the moment. Here's an example:

```python
import patito as pt
import polars as pl


class Struct(pt.Model):
    x: int
    y: int
    z: int = pt.Field(lt=2)


class MyModel(pt.Model):
    struct_col: Struct
    list_struct_col: list[Struct]


df = pl.DataFrame(
    {
        "struct_col": [{"x": 1, "y": 2, "z": 3}],
        "list_struct_col": [[{"x": 1, "y": 2, "z": 3}, {"x": 1, "y": 2, "z": 3}]],
    }
)

MyModel.validate(df)
```

This passes validation, despite all of the `z` fields being greater than 2. I expect to receive errors for both of the columns. Can `pt.Field` be improved to do more validation inside of structs?

### PR DESCRIPTION
**Pull request by dsgibbons**
Closes #43.

This fix takes advantage of the existing validation logic by recursively calling `_find_errors` whenever `pl.Struct` or `pl.List(pl.Struct)` columns are detected. 

This fix exposes a (possible) bug with one of the pre-existing test cases `test_model.test_missing_date_struct`. I've currently added a `@pytest.mark.skip` to the failing test. Is this actually a bug, and if so, can I continue to skip this test case and raise a new issue to fix it? Do you have any insights into why `Test.model_fields['c'].annotation.columns` does not exist?

### CODE REVIEW

**Comment by thomasaarholt** [comment]
I think this is a seriously elegant solution, and I really hope we can make it work for the general case.

I think the skipped test error is related to the following example:

Consider a nullable struct column:
```python
import patito as pt

class XY(pt.Model):
    x: int
    y: int

class Coord(pt.Model):
    id: int
    xy: Optional[XY] # <-- nullable
```

What we are saying here is that we allow the following values for `xy`:
- `{"x":1, y:1}`
- `null`

But not
- `{"x":1, y:null}`

Now let's take a look at how polars materializes the above three column values:
```python
import polars as pl

schema = {"id": pl.Int64, "xy": pl.Struct({"x": pl.Int64, "y": pl.Int64})}
# note the difference between the two last elements in the struct:
data = {"id": [0, 1, 2], "xy": [{"x": 1, "y": 1}, None, {"x": None, "y": None}]}

df = pl.DataFrame(data, schema=schema)
print(df)
---
shape: (3, 2)
┌─────┬─────────────┐
│ id  ┆ xy          │
│ --- ┆ ---         │
│ i64 ┆ struct[2]   │
╞═════╪═════════════╡
│ 0   ┆ {1,1}       │
│ 1   ┆ {null,null} │
│ 2   ┆ {null,null} │
└─────┴─────────────┘
``` 

I think I was expecting `null` for the second row, that is, looking only at the last two rows:
```python
shape: (3, 2)
┌─────┬─────────────┐
│ id  ┆ xy          │
│ --- ┆ ---         │
│ i64 ┆ struct[2]   │
╞═════╪═════════════╡
│ 1   ┆ null        │
│ 2   ┆ {null,null} │
└─────┴─────────────┘ 
```

I'm going to have to have a little think about how we do this one 😊

File: src/patito/validators.py [hunk 1]
```diff
@@ -256,6 +256,26 @@ def _find_errors(  # noqa: C901
                     )
                 )
 
+        # intercept struct columns, and get errors seperately
+        if schema.dtypes[column_name] == pl.Struct:
+            struct_errors = _find_errors(
+                dataframe=dataframe.select(column_name).unnest(column_name),
+                schema=schema.model_fields[column_name].annotation,
```
Comment (thomasaarholt) [inline src/patito/validators.py]: Ooh, the reason (or at least, the _first_ reason) that the test failing is actually simpler:
The type of `schema` passed in the struct case on line 263 _should_ be a patito class (of `Type[Model]`), but in this case it is an instance of `typing._UnionGenericAlias`, basically the union `Type[Model] | None`.

Here we would need to extract the model from the union, and _ideally_ also be able to pass the nullable requirement like I described in the above comment.

Comment (dsgibbons) [reply src/patito/validators.py]: Ok, so this part can be addressed by using a try-except on `...annotation.__args__[0]` (or `...annotation.__args__[0].__args__[0]` if it is an optional list of structs) - see d1c3a3b6a8b33889e49f92f1ad122e8a446c9aad. 

Now, I'm just getting some missing values issues, because I haven't passed back the nullable requirement. I've re-added the failing test. This nullable requirement is complicated by the fact that, if all columns are required in the submodel, then for a given row, *all* values need to be either null or not null.

Comment (dsgibbons) [reply src/patito/validators.py]: Perhaps, we could filter out rows in the submodel that have all-null columns before calling `_find_errors` again?

Comment (dsgibbons) [reply src/patito/validators.py]: Commit deb1493cdd24c11ba126209a95c729512583de51 filters out all-null rows via `dataframe.filter(pl.col(column_name).is_not_null())` before calling `_find_errors` again.

**Comment by dsgibbons** [comment]
@thomasaarholt I've updated the PR to correctly select `Type[Model]` when dealing with `Optional` and have applied filters on all-null subrows to encode the nullable requirement, as discussed above.

Since polars coerces `None -> {null, null, ...}` when dealing with structs, I think this fix is valid. The only way I can see a more "legitimate" way of implementing this would be if polars itself either:
- Provided  a `pl.Optional` datatype for marking structs as optional.
- Changed its default coercion logic for handling `None` in the context of structs. Like @thomasaarholt, I was surprised to discover how polars currently coerces `None` for structs, as every other datatype I have come across (including `pl.List`) is simply encoded as `null`... there must be some good reason for doing this that is unlikely to change.

I'll likely update this PR to include a few more test cases for testing optional lists of structs and optional deeply nested structs, but I don't think the current fix will need to change. What do you think?

**Comment by thomasaarholt** [comment]
Super! I overall agree with you - I've asked [here](https://discord.com/channels/908022250106667068/1014967651656814694/1226191957039124531) on the discord just to make sure the implementation is correct. This is kinda an edge case, but someone might have an opinion.

**Comment by chainyo** [comment]
Hi any news on this PR?

**Comment by thomasaarholt** [comment]
Let's go ahead and merge this, as well make a new release. @dsgibbons, if you want to make a few more test cases, especially demonstrating null behaviour, in another PR, then I'd appreciate it.

**Comment by dsgibbons** [comment]
Yes, I'm still looking at adding additional test cases @thomasaarholt. I've been a bit busy lately but hopefully I'll be able to add more test cases over the next few weeks. Thanks for accepting the PR!
